### DISCUSSION TITLE
What should be the defaults?

### DISCUSSION THREAD

**Discussion by nojhan** [liquidprompt/liquidprompt#802] [General]
Since Liquid Prompt is advancing toward its version 2.2, I am wondering what should we use as the default configuration.

The default theme is obvious, but I was rather thinking of the colors (and icons set?).
So far, we use an 16-colors scheme, which shows its limits with so many features.
But we have 256-colors preset(s), that look quite good and would be a good option for a default.

I think that most of users now should have a terminal emulator supporting 256-colors,
so that it would make sense to have the default be a good looking one, and allow to fall back to 16-colors in case of problems.

Any ideas?

**Comment by Rycieos** [comment]
Colors are a pretty easy thing to configure, so I think it is fine if we change the defaults.

I think we should export the current defaults as a preset, to make switching back easy if any user hates the change.

I will let you pick new colors because of my vision limitations. But keep the contrast between different states of the same section high.

**Comment by nojhan** [comment]
I started a (WIP) branch, with a [tool to check colors contrast](https://github.com/liquidprompt/liquidprompt/blob/refactor/default-theme/tools/color-contrast.sh), it follows the W3C's recommended levels for colorblindness-friendly color choice. The existing themes may have to be improved if we want to have a AAA level of contrast for the default.

![color-contrast_2](https://github.com/liquidprompt/liquidprompt/assets/285009/442416cf-3e7c-4d0e-ab52-a537886e94e2)

I am also wondering what is the most important contrast:
1. between sections (e.g. the path section VS the devtools one), then states (e.g. random path element VS VCS repo element)?
2. between states, then sections?

**Reply by Rycieos** [reply]
That is a pretty cool tool. It needs some explanation of both how it works and how it is supposed to be used. It seems that it will show colors that are high enough contrast from the input color?

As to your question, I say the contrast between states is most important. Sections are pretty obviously separated by other visual cues, but some information is *only* communicated by color, like VCS branch color communicating dirty, ahead, or matching.

**Reply by nojhan** [reply]
Yes, if you pass one ANSI color as an argument, it will show you all colors that are contrasting enough (as per the W3C definition, with two levels of contrasts). If you pass two colors, it will print the contrast level and exit on error if it is below AAA. If you call it without argument, it prints the ANSI tables.

**Reply by nojhan** [reply]
Here are my current thoughts, after reading some more literature and making some tests:

- There is no color in ANSI 256 with a AAA contrast level with both black and white. Which makes it impossible to design a theme that would work for both light and dark terminal background.
- There is only a few colors there with a AA contrast level with both black and white.
- Note of those colors have an AA contrast level between them.
- It is thus virtually impossible to find colors that would work for against any kind of background, have sufficient contrast between states AND sufficient contrast between segments.
- We may release the background constraint by considering only ANSI 16, as most terminals allow to adjust those colors.
- The least problematic way is to have very few colors in the prompt: the default (I assume white or black), one for notes and another for warnings. That way, contrast is maximized between states, and we rely on the position for discriminating segments.
- If we want to honor all those constraints, that leaves us with only green, yellow and cyan. Among which cyan/yellow is probably the best option for targeting a majority of people.

I tried to implement that in the ["few-colors" preset](https://github.com/liquidprompt/liquidprompt/blob/refactor/default-theme/contrib/presets/colors/few-colors.conf). But of course it looks spartiate and my guess is that it would make a bad default for naive users.

<img width="821" alt="Screenshot 2023-11-19 at 20 29 25" src="https://github.com/liquidprompt/liquidprompt/assets/285009/22cb5a26-da35-4ca5-996f-a05edd880430">

**Reply by Rycieos** [reply]
I agree, it would make a bad default. My biggest issue with it is that most things are bland (though I'm sure _someone_ would like that). In my opinion, I only need colors that are high contrast when one piece of the prompt can have multiple colors based on some state. For example, the CPU temperature. I need to be able to differentiate between cool, warm, and hot. The contrast between lines added, lines subtracted, commits ahead, and commits behind is not important to me, since every time those appear, they are always the same color.

**Reply by nojhan** [reply]
Yeah. I actually got some people commenting on [the (french) comparison article](https://linuxfr.org/news/comparaison-critique-de-systemes-d-invite-de-commande) and after [my (french) presentation](https://cfp.capitoledulibre.org/cdl-2023/talk/KYLR9C/) with just that kind of comment (wanting some "minimalism").

As for the default, I am now rather leaning on keeping the current default (which works on dark/light background and has some contrast, I will see if it can be improved a bit), but emphasize more in the documentation that the user can easily have another configuration.
